### PR TITLE
Update cmd_sync

### PR DESCRIPTION
**Pull request by ashleydavies**
Verified locally; fetches & deletes files as expected, including if entire directories are deleted.

Extra bonus is we don't need the special case for an empty local directory anymore, since every file will naturally be fetched :tada:

Co-authored-by: Megan Sumner <megansumner@improbable.io>

### CODE REVIEW

**Comment by ashleydavies** [comment]
Just discovered that this doesn't delete unexpected directories; I'll investigate why that is tomorrow and see if I can get it working for that too

**Comment by ashleydavies** [comment]
Now:
* Deletes folders which are newly empty from deleting all of the files within them
  * This is a compromise - there doesn't seem to be a nice way to find all folders locally that aren't on the remote, so we assume folders the remote doesn't tell us about that we deleted items from should be deleted
* Requires `--delete` to delete files; otherwise just outputs what it would like to delete
* Doesn't make blank snapshots on the remote server (... although this is a bit of an ugly hack - extracted a `DrySnapshot` method with the bulk of the `Snapshot` logic)

Here's a list of times to checkout a bunch of consecutive nightly builds:

```
       60.77 real        94.85 user         8.66 sys
       59.96 real        89.50 user         8.82 sys
       56.37 real        82.97 user         7.64 sys
       65.03 real        92.21 user        10.01 sys
       59.41 real        88.46 user         8.33 sys
       97.21 real       116.98 user        17.14 sys
       56.70 real        83.73 user         7.56 sys
       74.70 real        96.70 user        11.30 sys
       67.47 real        99.09 user        11.95 sys
       59.68 real       100.76 user         9.04 sys
       63.19 real        97.94 user        10.02 sys
       72.36 real       103.24 user        12.12 sys
       95.36 real       116.90 user        17.42 sys
       67.95 real       106.95 user        12.43 sys
       78.35 real       104.11 user        13.61 sys
       58.93 real        84.51 user         8.24 sys
       72.42 real        94.60 user        10.87 sys
       50.84 real        71.81 user         7.68 sys
```

There's some spikes, presumably with big content changes, but it seems steady at ~60s (previously it takes ~150s on my machine)

Measured with:
```fish
for checkout in (../restic/restic -r /Volumes/clientcache/ snapshots | grep cloud-slave | grep nightly | awk '{print $1}')
    time ../restic/restic -r /Volumes/clientcache/ sync $checkout --target . --no-lock --progress --delete >../output.log
end
```

**Comment by ashleydavies** [comment]
@ScottBrooks I think this is ready for a review now 🎉

File: cmd/restic/cmd_sync.go [hunk 1]
```diff
@@ -99,7 +112,8 @@ func (ft *FileTracker) CompleteItem(item string, previous, current *restic.Node,
 func runSync(opts SyncOptions, gopts GlobalOptions, term *termstatus.Terminal, args []string) error {
 	ctx := gopts.ctx
 
-	fileTracker := FileTracker{}
+	start := time.Now()
```
Comment (ScottBrooks) [inline cmd/restic/cmd_sync.go]: We can probably drop these time.Now()  trackers that we were using to record the time various steps took.

File: cmd/restic/cmd_sync.go [hunk 2]
```diff
@@ -177,11 +189,13 @@ func runSync(opts SyncOptions, gopts GlobalOptions, term *termstatus.Terminal, a
...
 	}
-	scanResult := func(item string, s archiver.ScanStats) {
-		Printf("Item: %s Stats: %+v\n", item, s)
 
+	scanResult := func(item string, s archiver.ScanStats) {
+		//Printf("Item: %s Stats: %+v\n", item, s)
 	}
 
+	preScan := time.Now()
+	Verbosef("Time to start scanning: %s\n", preScan.Sub(start))
```
Comment (ScottBrooks) [inline cmd/restic/cmd_sync.go]: We can kill these various timestamp logs.

File: cmd/restic/cmd_sync.go [hunk 3]
```diff
@@ -231,24 +251,30 @@ func runSync(opts SyncOptions, gopts GlobalOptions, term *termstatus.Terminal, a
...
-				Warnf("error for include pattern: %v", err)
-			}
+	res.SelectFilter = func(item string, dstpath string, node *restic.Node) (selectedForRestore bool, childMayBeSelected bool) {
+		swapped := strings.Replace(item, "\\", "/", -1)
+		shouldGet, ok := fileTracker.FilesToGet[swapped]
 
-			selectedForRestore = matched
-			childMayBeSelected = (childMayMatch) && node.Type == "dir"
-			Printf("Restoring: %s %s %v\n", item, dstpath, matched)
+		//matched, childMayMatch, err := filter.List(fileTracker.FilesToGet, item)
```
Comment (ScottBrooks) [inline cmd/restic/cmd_sync.go]: We can remove this commented out line as well.

File: cmd/restic/cmd_sync.go [hunk 4]
```diff
@@ -269,9 +295,69 @@ func runSync(opts SyncOptions, gopts GlobalOptions, term *termstatus.Terminal, a
...
+			}
+		}
+		Verbosef("Files Fetched: %v\n", filesFetched)
+		Verbosef("Files To Delete: %v\n", fileTracker.FilesToDelete)
+		Verbosef("Directories To Delete: %v\n", fileTracker.UnexpectedDirs)
+		Verbosef("Total time: %s\n", time.Now())
+	}
+
+	if opts.Delete {
+		Verbosef("%d errors while deleting files\n", deleteFilesAndDirectories(fileTracker, opts))
```
Comment (ScottBrooks) [inline cmd/restic/cmd_sync.go]: Can we put the deleteFilesAndDirectories on it's own line and then print out the amount?  That way if we are commenting out some logging, we don't loose the function call as well.

**Review by ScottBrooks — CHANGES_REQUESTED** [review]
In order to make it as easy as possible to rebase this on top of restic, can I get you to roll back any unnecessary changes to the core restic stuff?

I know it's asking for  worse code, but it's going to be a real pain to integrate changes.

File: cmd/restic/cmd_sync.go (hunk 4 above)
Comment (ashleydavies) [reply cmd/restic/cmd_sync.go]: Good spot, thanks!

**Comment by ashleydavies** [comment]
@ScottBrooks All done - only additive now, so should be straightforward to rebase. I wanted to move `DrySnapshot` into cmd_sync, but it makes use of other private functions in `archiver.go`

File: cmd/restic/cmd_sync.go [hunk 5]
```diff
@@ -269,9 +282,67 @@ func runSync(opts SyncOptions, gopts GlobalOptions, term *termstatus.Terminal, a
...
+		if err != nil {
+			errorCount++
+			Printf("Error deleting %s\n", fileName)
+		}
+	}
+
+	// As we are not using RemoveAll (hence why we do files first), we cannot delete non-empty directories, so a bug in
+	// the directory marking logic cannot result in the deletion of directories containing files
+	// We must delete deeper directories first, however - e.g. in the case of /A/B/, we must first delete B/, then A/
+	depthSortedUnexpectedDirs := map[int][]string{}
```
Comment (ScottBrooks) [inline cmd/restic/cmd_sync.go:325]: This is a nice implementation of deleting our paths with the most components first.

**Review by ScottBrooks — APPROVED** [review]
Looks good to me, nice work.

File: cmd/restic/cmd_sync.go (hunk 5 above)
Comment (ashleydavies) [reply cmd/restic/cmd_sync.go:325]: Thanks! I'm a little scared there may be some crazy edge case if e.g. directory symlinks are traversed, but I figure that shouldn't be the case. Worst case we'll just have a few leftover empty directories with errors reported in the update log

**Pull request merged**